### PR TITLE
Enable RRef timeout for tensorpipe

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39531 Enable RRef timeout for tensorpipe**
* #39530 Cleanup rref_impl

Enables RRef timeout support in TP agent by having TP agent mark
timeout errors with `makeRPCError` API. Also does some refactoring so TP agent
can print out the timeout for each future that has timed out.

Differential Revision: [D21881475](https://our.internmc.facebook.com/intern/diff/D21881475/)

### CODE REVIEW

**Review by mrshenli — APPROVED** [review]
LGTM! Please get a stamp from @lw too

**Comment by rohan-varma** [comment]
@lw Do you mind taking a look at this diff? Thanks!

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 1]
```diff
@@ -23,6 +24,8 @@ const std::string kNumIdleThreads = "agent.num_idle_threads";
 const std::string kClientActiveCalls = "agent.client_active_calls";
 const std::string kServerActiveCalls = "agent.server_active_calls";
 const std::string kServerActiveAsyncCalls = "agent.server_active_async_calls";
+const std::string kTensorPipeRPCTimeoutErrorStr =
```
Comment (osalpekar) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:27]: nit: kTensorPipeRpcTimeoutErrorStr

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 2]
```diff
@@ -418,7 +421,8 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
     {
       std::unique_lock<std::mutex> lock(timeoutMapMutex_);
       auto& timeoutFuturesVector = timeoutMap_[expirationTime];
-      timeoutFuturesVector.emplace_back(futureResponseMessage);
+      timeoutFuturesVector.emplace_back(
+          std::make_pair(futureResponseMessage, timeout));
```
Comment (osalpekar) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:425]: I think you might not need to call `std::make_pair` with `emplace_back`?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 3]
```diff
@@ -2833,7 +2833,6 @@ def test_return_future(self):
                 return_future,
             )
 
-    @_skip_if_tensorpipe_agent
     @dist_init
```
Comment (osalpekar) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:2836]: Has this been stress-tested? Just checking to ensure we don't introduce more flaky TensorPipe agent tests

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp (hunk 1 above)
Comment (lw) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:27]: nit: no need for the `TensorPipe` prefix, this is in anonymous namespace, it's local to this file

**Review by lw — APPROVED** [review]
Looks great, thanks!

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp (hunk 2 above)
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:425]: Great catch!

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 3 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:2836]: Ran it with 500 stresstests and they all passed.

**Pull request closed without merging**